src/crypto/CHIPCryptoPALmbedTLS.cpp

```diff
@@ -1405,39 +1405,62 @@ CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan
     return ExtractKIDFromX509Cert(false, certificate, akid);
 }
 
-CHIP_ERROR ExtractVIDFromX509Cert(const ByteSpan & certificate, VendorId & vid)
+namespace {
+
+CHIP_ERROR ExtractDNAttributeFromX509Cert(bool isVID, const ByteSpan & certificate, uint16_t & id)
 {
 #if defined(MBEDTLS_X509_CRT_PARSE_C)
     CHIP_ERROR error = CHIP_NO_ERROR;
     mbedtls_x509_crt mbed_cert;
+    mbedtls_asn1_named_data * asn1IdEntry;
 
     mbedtls_x509_crt_init(&mbed_cert);
 
     int result = mbedtls_x509_crt_parse(&mbed_cert, Uint8::to_const_uchar(certificate.data()), certificate.size());
     VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
 
     // vendor id is the second element of subject and should be of size 4
+    // product id is the third element of subject and should be of size 4
     // returning CHIP_ERROR_KEY_NOT_FOUND to sinalize VID is not present in the certificate.
-    VerifyOrExit(mbed_cert.subject.next->val.p != nullptr && mbed_cert.subject.next->val.len == 4,
+    asn1IdEntry = isVID ? mbed_cert.subject.next : mbed_cert.subject.next->next;
+    VerifyOrExit(asn1IdEntry != nullptr && asn1IdEntry->val.p != nullptr && asn1IdEntry->val.len == 4,
                  error = CHIP_ERROR_KEY_NOT_FOUND);
 
     // vendor id is of size 4, we should ensure the string is null terminated before passing in to strtoul to avoid undefined
     // behavior
-    mbed_cert.subject.next->val.p[4] = 0;
-    vid = static_cast<VendorId>(strtoul(reinterpret_cast<const char *>(mbed_cert.subject.next->val.p), NULL, 16));
+    asn1IdEntry->val.p[4] = 0;
```

Comment:
This mutates the buffer in an unsafe way. We should not mutate the buffer. We should instead:

1- Check we have expected length (already done)
2- Create a temporary stack buffer of length (4 + 1)
3- Copy 4 bytes from asn1IdEntry->val.p into temporary stack buffer
4- Null-terminate stack buffer
5- Convert stack buffer

This avoids possibly corrupting the data in the parsed cert that would make it dangerous to eventually modify this method to do more checks.